<p align="center">
<img src="https://pypi.org/static/images/logo-small.8998e9d1.svg" width="" height="">
</p>

# **<font color="#07a8ed">Estadística descriptiva con Python**

## **<font color="#07a8ed">Librerías**

In [1]:
!pip install yfinance

https://pypi.org/project/yfinance/


In [2]:
import pandas as pd
import numpy as np
import scipy.stats as ss
import yfinance as yf
import plotly.express as px
import plotly.graph_objects as go

## **<font color="#07a8ed">Obtención de datos**


<p align="justify">
Con `yfinance` accedemos a la base de yahoo finance y le indicanmos en nombre del ticker que queremos analizar, en este caso el activo seleccionado será el de Apple Inc. (AAPL). Si indicamos <code>['Adj Close']</code> nos descargará los precios de cierre ajustados del activo.
<br><br>
La fecha es también modificable y podemos añadir el parámetro `end` para delimitar la finalización. En este caso, no hemos indicado fecha de cierre, por lo que toma la fecha actual.

In [3]:
ticker = ['AAPL']
aapl = yf.download(ticker, start= "2015-1-1", interval='1d', rounding='true')[['Adj Close']].copy()
aapl.head()

[*********************100%***********************]  1 of 1 completed


Adj Close
Date                 
2015-01-02      24.37
2015-01-05      23.69
2015-01-06      23.69
2015-01-07      24.02
2015-01-08      24.94

<p align="justify">
Generamos una nueva columna en nuestro dataframe que contenga los rendimientos logarítmicos.

In [4]:
aapl["returns"] = np.log(aapl["Adj Close"]/aapl["Adj Close"].shift(1)) #método shift de pandas desplaza el índice del dataframe.

In [19]:
aapl.head()

Date  Adj Close   returns
0 2015-01-05      23.69 -0.028300
1 2015-01-06      23.69  0.000000
2 2015-01-07      24.02  0.013834
3 2015-01-08      24.94  0.037586
4 2015-01-09      24.97  0.001202

In [34]:
import numpy as np  ## metodo para  calcular el componente de la formula de logaritmo primer dato

x = np.exp(-0.02830) * 23.69
print(x)

23.028970681909648


In [36]:
np.log( 23.0289/23.69000) ## calcular x



-0.028303069265191886

In [30]:
np.log(23.69/23.69)

0.0

In [29]:
np.log(24.94/24.02)  ## logaritmo de la posicion

0.03758612360005302

In [37]:
aapl.dropna(inplace=True)

In [38]:
aapl.head()

Date  Adj Close   returns
0 2015-01-05      23.69 -0.028300
1 2015-01-06      23.69  0.000000
2 2015-01-07      24.02  0.013834
3 2015-01-08      24.94  0.037586
4 2015-01-09      24.97  0.001202

## **<font color="#07a8ed">Cálculo de estadísticos y visualización de los datos**

### **<font color="#07a8ed">Precio**

In [8]:
precio = aapl["Adj Close"]

cantidad_datos = len(precio)
media_precio = np.mean(precio)
mediana_precio = np.median(precio)
desviacion_precio = np.std(precio)
varianza_precio = np.var(precio)
asimetria_precio = ss.skew(precio)
curtosis_precio = ss.kurtosis(precio)
maximo_precio = np.max(precio)
minimo_precio = np.min(precio)
rango_precio = maximo_precio - minimo_precio
q1 = np.quantile(precio, 0.25)
q2 = np.quantile(precio, 0.5)
q3 = np.quantile(precio, 0.75)

print(f"""
Estadísticos del precio de Apple Inc.

    Cantidad de datos: {cantidad_datos}
    Media: {media_precio:.2f}
    Mediana: {mediana_precio:.2f}
    Desviación estándar: {desviacion_precio:.2f}
    Varianza: {varianza_precio:.2f}
    Coeficiente de asimetría: {asimetria_precio:.2f}
    Curtosis: {curtosis_precio:.2f}
    Precio máximo: {maximo_precio}
    Precio mínimo: {minimo_precio}
    Rango: {rango_precio}
    Primer cuartil: {q1}
    Segundo cuartil: {q2}
    Tercer cuartil: {q3}
""")


Estadísticos del precio de Apple Inc.

    Cantidad de datos: 2450
    Media: 90.66
    Mediana: 60.22
    Desviación estándar: 62.42
    Varianza: 3896.57
    Coeficiente de asimetría: 0.47
    Curtosis: -1.28
    Precio máximo: 234.55
    Precio mínimo: 20.72
    Rango: 213.83
    Primer cuartil: 34.7025
    Segundo cuartil: 60.215
    Tercer cuartil: 147.7175



<p align="justify">
Según los valores obtenidos, se observa que la media es de 90.66, con una desviación estándar de 62.42; lo cual ubica los precios entre 23.03 y 153.08  (con un 68% de confianza). El Coeficiente de Asimetría (Skewness) de 0.47, indica que hay asimetría a la izquierda de la distribución (media mayor a mediana). El Coeficiente de Curtosis (Kurtosis) menor a cero, indica que la distribución es platicúrtica, es decir, hay poca concentración de datos en la media.
<br><br>
Además, realizando un análisis de cuartiles se concluye que el 25% de los valores están por debajo de 34.70, el 50% por debajo de 60.22 (mediana) y el 75% por debajo de 147.71%.

Con `describe` de Pandas:

In [9]:
precio.describe()

count    2450.000000
mean       90.664494
std        62.435255
min        20.720000
25%        34.702500
50%        60.215000
75%       147.717500
max       234.550000
Name: Adj Close, dtype: float64

In [10]:
px.histogram(x=precio,
             template="gridon",
             title = "Histograma",
             labels={"x":"rendimiento"}).update_layout(bargap=0.2)

In [11]:
aapl.head()

Adj Close   returns
Date                           
2015-01-05      23.69 -0.028300
2015-01-06      23.69  0.000000
2015-01-07      24.02  0.013834
2015-01-08      24.94  0.037586
2015-01-09      24.97  0.001202

In [12]:
fig = px.line(aapl,
              x=aapl.index,
              y="Adj Close",
              template="gridon",
              labels={"Date":"Fecha","Adj Close":"Precio de cierre"}
)
fig.show()

In [13]:
aapl = aapl.reset_index()
aapl.head()

Date  Adj Close   returns
0 2015-01-05      23.69 -0.028300
1 2015-01-06      23.69  0.000000
2 2015-01-07      24.02  0.013834
3 2015-01-08      24.94  0.037586
4 2015-01-09      24.97  0.001202

In [14]:
x0 = aapl["Date"].iloc[0]               # La primera
x1 = aapl["Date"].iloc[-1]              # La última
xm = aapl["Date"].iloc[len(aapl)//2]    # La del medio

fig = px.line(aapl,
              x="Date",
              y="Adj Close",
              template="gridon",
              labels={"Date":"Fecha","Adj Close":"Precio de cierre"}
)

#la linea roja:
fig.add_shape(type="line",
              x0=x0, y0=media_precio, x1=x1, y1=media_precio,
              line=dict(color="Red",width=2)
 )

#el texto 10 píxeles por encima de la línea:
fig.add_annotation(x=xm, y=media_precio,
                   text='Precio promedio',
                   showarrow=False,
                   yshift=10)

### **<font color="#07a8ed">Rendimiento diario**

In [15]:
rendimiento = aapl["returns"]

media_rend = np.mean(rendimiento)
mediana_rend = np.median(rendimiento)
desviacion_rend = np.std(rendimiento)
varianza_rend = np.var(rendimiento)
asimetria_rend = ss.skew(rendimiento)
curtosis_rend = ss.kurtosis(rendimiento)

print(f"""
Estadísticos del rendimiento mensual de Apple Inc.

    Media: {media_rend*100:.2f} %
    Mediana: {mediana_rend*100:.2f} %
    Desviación estándar: {desviacion_rend*100:.2f} %
    Coeficiente de asimetría: {asimetria_rend:.2f}
    Curtosis: {curtosis_rend:.2f}
""")


Estadísticos del rendimiento mensual de Apple Inc.

    Media: 0.09 %
    Mediana: 0.09 %
    Desviación estándar: 1.81 %
    Coeficiente de asimetría: -0.20
    Curtosis: 5.39



In [40]:
px.histogram(x=rendimiento,
             template="gridon",
             title = "Histograma",
             labels={"x":"rendimiento"}).update_layout(bargap=0.1)


In [ ]:
fig = px.line(aapl,
              x="Date",
              y="returns",
              template="gridon",
              labels={"Date":"Fecha","returns":"rendimiento"}
)

#la linea roja (media):
fig.add_shape(type="line",
              x0=x0, y0=media_rend, x1=x1, y1=media_rend,
              line=dict(color="Red",width=2)
)

#las lineas naranjas (desviación estándar):
fig.add_shape(type="line",
              x0=x0, y0=media_rend+desviacion_rend, x1=x1, y1=media_rend+desviacion_rend,
              line=dict(color="orange",width=3)
)
fig.add_shape(type="line",
              x0=x0, y0=media_rend-desviacion_rend, x1=x1, y1=media_rend-desviacion_rend,
              line=dict(color="orange",width=3)
)
